### PR TITLE
Response body bounds

### PR DESCRIPTION
**Pull request by jack-berg**
Add new max response body size parameter to sender config, defaulting to 4mb.

Per the spec, if a response exceeding the limit is received:
- grpc: fail with RESOURCE_EXHAUSTED code, do not retry
- http: fail with exception, do not retry

For OTLP exporters, the body size is not configurable, since we don't do anything with the response today. If adding this 4mb limit causes new exceptions for users, we can add a config option to allow it to be increased.

For JaegerRemoteSampler, a new programmatic config option is added because the responses are critical to the function of the sampler. 

cc @brunobat you'll want to update the vertex sender to incorporate this new option.

Proto spec update: https://github.com/open-telemetry/opentelemetry-proto/pull/781

### CODE REVIEW

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java [hunk 1]
```diff
@@ -230,7 +230,9 @@ public GrpcExporter build() {
                 isPlainHttp ? null : tlsConfigHelper.getSslContext(),
                 isPlainHttp ? null : tlsConfigHelper.getTrustManager(),
                 executorService,
-                grpcChannel));
+                grpcChannel,
+                // 1kb since don't do anything with responses today
+                1024L));
```
Comment (jack-berg) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java]: TODO: Review this default in the context of https://github.com/open-telemetry/opentelemetry-proto/pull/781/changes

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java [hunk 2]
```diff
@@ -230,7 +230,10 @@ public GrpcExporter build() {
                 isPlainHttp ? null : tlsConfigHelper.getSslContext(),
                 isPlainHttp ? null : tlsConfigHelper.getTrustManager(),
                 executorService,
-                grpcChannel));
+                grpcChannel,
+                // 4mb to align with spec guidance - even though we don't do anything with the
```
Comment (zeitlinger) [inline exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java:234]: Issue says 1mb

File: exporters/sender/jdk/src/test/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSenderTest.java [hunk 3]
```diff
@@ -66,7 +66,8 @@ void setup() throws IOException, InterruptedException {
             Duration.ofSeconds(10),
             Collections::emptyMap,
             RetryPolicy.builder().setMaxAttempts(2).setInitialBackoff(Duration.ofMillis(1)).build(),
-            null);
+            null,
+            Long.MAX_VALUE);
```
Comment (zeitlinger) [inline exporters/sender/jdk/src/test/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSenderTest.java:70]: Test coverage is entirely absent for the feature

Every test and benchmark passes Long.MAX_VALUE, which bypasses the bound completely. There are no tests for:
  - A response body that exceeds maxResponseBodySize (truncation at limit)
  - Boundary case (exactly at limit)
  - What happens with very small values

  This is the most important thing the response limit is supposed to do — it should have at least one test per sender proving truncation works.

File: exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java [hunk 4]
```diff
@@ -82,7 +82,8 @@ public final class JdkHttpSender implements HttpSender {
       Duration timeout,
       Supplier<Map<String, List<String>>> headerSupplier,
       @Nullable RetryPolicy retryPolicy,
-      @Nullable ExecutorService executorService) {
+      @Nullable ExecutorService executorService,
+      long maxResponseBodySize) {
```
Comment (zeitlinger) [inline exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java:87]: JaegerRemoteSamplerBuilder.setMaxSamplingStrategyResponseBodySize validates bytes > 0. But the sender constructors (JdkHttpSender, OkHttpHttpSender, OkHttpGrpcSender) accept any long without
  validation. A caller can bypass the guard by constructing a sender directly with 0 or -1. Consider adding the validation at the sender level too, or document the expected invariant.

File: exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpHttpSender.java [hunk 5]
```diff
@@ -160,7 +164,16 @@ public String getStatusMessage() {
                                 public byte[] getResponseBody() {
                                   if (bodyBytes == null) {
                                     try {
-                                      bodyBytes = body.bytes();
+                                      Buffer buffer = new Buffer();
+                                      while (buffer.size() < maxResponseBodySize) {
```
Comment (zeitlinger) [inline exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpHttpSender.java]: In both OkHttpHttpSender and OkHttpGrpcSender, after the loop exits at the limit, the remaining body bytes aren't consumed. OkHttp can only return a connection to the pool if the body is fully consumed or explicitly discarded (via body.close()). The callers do close the response, but closing without full consumption causes OkHttp to close the TCP connection rather than pool it. For the (presumably rare) case where a server sends an oversized body, this means a new connection on the next request. Probably acceptable, but worth a comment explaining the tradeoff.

File: exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java [hunk 6]
```diff
@@ -273,21 +275,16 @@ HttpResponse<byte[]> sendInternal(MessageWriter requestBodyWriter) throws IOExce
     throw exception;
   }
 
-  private static String responseStringRepresentation(HttpResponse<?> response) {
-    StringJoiner joiner = new StringJoiner(",", "HttpResponse{", "}");
-    joiner.add("code=" + response.statusCode());
-    joiner.add(
-        "headers="
```
Comment (zeitlinger) [inline exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java:280]: why is that dropped?

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerBuilder.java [hunk 7]
```diff
@@ -147,6 +149,16 @@ public JaegerRemoteSamplerBuilder setInitialSampler(Sampler initialSampler) {
     return this;
   }
 
+  /**
+   * Sets the maximum number of bytes to read from a sampling strategy response body. If unset,
+   * defaults to 4 MiB. Must be positive.
+   */
+  public JaegerRemoteSamplerBuilder setMaxSamplingStrategyResponseBodySize(long bytes) {
```
Comment (zeitlinger) [inline sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerBuilder.java:156]: GrpcExporterBuilder and HttpExporterBuilder hardcode 4MB with no setter. Only JaegerRemoteSamplerBuilder exposes a public setter. If a user wants to tighten the OTLP limit (say, to 64KB since responses are tiny), they can't. Worth a deliberate decision: intentionally omit setters for OTLP (since responses are well-known small), or add them for consistency?

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java (hunk 2 above)
Comment (martincostello) [reply exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java:234]: Issue is just outdated. 4MB is the correct value: https://github.com/open-telemetry/opentelemetry-proto/pull/781

File: sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerBuilder.java (hunk 7 above)
Comment (jack-berg) [reply sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/JaegerRemoteSamplerBuilder.java:156]: I omitted them in OTLP for now because we don't do anything with them and a 4mb limit is an improvement on the current no limit.

In contrast, for JaegerRemoteSampler, the default body size reduction is a change in behavior which could be meaningful so adding the setter seems essential.

File: exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java (hunk 6 above)
Comment (jack-berg) [reply exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java:280]: There's a bug in java 11 http client inputstream handling that causes deadlocks, which is exercised when you try to call InputStream#readNBytes. The fix is in java 13+ and for some reason not back ported to java 11. Working around it using the existing code organization of passing around jdk HttpResponse<byte[]> is awkward. So I adjusted to pass around the otel HttpResponse instead. But as a result, we don't have access to headers for logging anymore.

(Can provide more details on the big when I'm back at my keyboard)

File: exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpHttpSender.java (hunk 5 above)
Comment (jack-berg) [reply exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpHttpSender.java]: Good callout. I will fix.

File: exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java (hunk 2 above)
Comment (jack-berg) [reply exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/GrpcExporterBuilder.java:234]: Ill update the pr description to match the latest

File: exporters/sender/jdk/src/test/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSenderTest.java (hunk 3 above)
Comment (jack-berg) [reply exporters/sender/jdk/src/test/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSenderTest.java:70]: I added test coverage, which was pretty tricky. In the process, was forced to do a more thorough review of the spec and updated the implementation to better match. Previously, when a payload exceeded the limit, I simply was truncating it and returning GrpcResponse / HttpResponse. Updated to return unretryable error instead per the spec (see PR description for details).

> Boundary case (exactly at limit)

I skipped testing at the boundary because its tricky to produce protobuf messages of an exact target size.

> it should have at least one test per sender proving truncation works.

In the updated implementation, there's no truncation. Instead an error response is returned, which is a convenient signal to assert against in tests.

File: exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java (hunk 4 above)
Comment (jack-berg) [reply exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java:87]: All the sender constructors are internal, and have a bunch of other unvalidated parameters which could be equally corrupted if a user goes around the guards. There's a conversation [here](https://github.com/open-telemetry/opentelemetry-java/issues/8173#issuecomment-4156480920) discussing where to add additional null checks beyond the guarantees from nullaway. I think we should adopt a policy of adding additional null checks at the API boundaries, but trusting nullaway once we're within the walled garden of internal code. The same would apply to parameter validation, I think.

**Comment by jack-berg** [comment]
I had an oversight where I was enforcing the size limit against the response pre-decompression.

This means that a response payload at the limit of 4mb, could decompress to a much larger size and have a big impact on memory.

Fixing this means more complexity, including new enforcement that we only support gzip compressed resposnes. This issue has turned in quite the rabbit hole.

On the bright side, there's pretty good abstract tests in AbstractHttpTelemetryExporterTest / AbstractGrpcTelemetryExporterTest that verify that all the sender implementations are doing the right thing.

File: exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java (hunk 4 above)
Comment (brunobat) [reply exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java:87]: Well... setting directly to a low value like -1 or 0 will effectively bloc data export...
see:
```
  if (bodyBytes.length > maxResponseBodySize) {
    throw new ResponseBodyTooLargeException(
        "HTTP response body exceeded limit of " + maxResponseBodySize + " bytes");
  }
```
All bodies will be bigger that that.

**Comment by brunobat** [comment]
Thanks @jack-berg.
About the config. We don't really use the `HttpSenderConfig` I'll have to map it to something like: `quarkus.otel.exporter.otlp.maxResponseBodySize`.

File: exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/ImmutableHttpResponse.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import com.google.auto.value.AutoValue;
+import io.opentelemetry.sdk.common.export.HttpResponse;
+
+@AutoValue
+abstract class ImmutableHttpResponse implements HttpResponse {
+
+  static ImmutableHttpResponse create(int statusCode, String statusMessage, byte[] responseBody) {
+    return new AutoValue_ImmutableHttpResponse(statusCode, statusMessage, responseBody);
+  }
+
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/ImmutableHttpResponse.java:17]: This is wrong. It gets `getStatusCode()` and `getStatusMessage` from the parent interface.

File: exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/ImmutableGrpcResponse.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+import javax.annotation.Nullable;
+
+@AutoValue
+abstract class ImmutableGrpcResponse implements GrpcResponse {
+
+  static ImmutableGrpcResponse create(
+      GrpcStatusCode statusCode, @Nullable String statusDescription, byte[] responseMessage) {
+    return new AutoValue_ImmutableGrpcResponse(statusCode, statusDescription, responseMessage);
+  }
+
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/ImmutableGrpcResponse.java:20]: This is wrong. It gets `getStatusCode()` and `getStatusMessage` from the parent interface.

File: exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/ImmutableHttpResponse.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import com.google.auto.value.AutoValue;
+import io.opentelemetry.sdk.common.export.HttpResponse;
+
+@AutoValue
+abstract class ImmutableHttpResponse implements HttpResponse {
+
+  static ImmutableHttpResponse create(int statusCode, String statusMessage, byte[] responseBody) {
+    return new AutoValue_ImmutableHttpResponse(statusCode, statusMessage, responseBody);
+  }
+
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/ImmutableHttpResponse.java:17]: This is wrong. It gets `getStatusCode()` and `getStatusMessage` from the parent interface.

File: exporters/sender/grpc-managed-channel/src/main/java/io/opentelemetry/exporter/sender/grpc/managedchannel/internal/UpstreamGrpcSender.java [hunk 11]
```diff
@@ -21,7 +21,6 @@
 import io.grpc.StatusRuntimeException;
 import io.grpc.stub.ClientCalls;
 import io.grpc.stub.MetadataUtils;
-import io.opentelemetry.exporter.internal.grpc.ImmutableGrpcResponse;
 import io.opentelemetry.exporter.internal.grpc.MarshalerInputStream;
 import io.opentelemetry.sdk.common.CompletableResultCode;
 import io.opentelemetry.sdk.common.export.GrpcResponse;
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/sender/grpc-managed-channel/src/main/java/io/opentelemetry/exporter/sender/grpc/managedchannel/internal/UpstreamGrpcSender.java:26]: This is incorrect. Internally, the grpc-java takes care of ensuring that oversized compressed and decompressed payloads don't reach RESPONSE_MARSHALER.parse in the following places:

https://github.com/grpc/grpc-java/blob/eac9fe9612881d54d51ddc5496599ba9a8201021/core/src/main/java/io/grpc/internal/MessageDeframer.java#L387

https://github.com/grpc/grpc-java/blob/eac9fe9612881d54d51ddc5496599ba9a8201021/core/src/main/java/io/grpc/internal/MessageDeframer.java#L445-L524

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractHttpTelemetryExporterTest.java [hunk 12]
```diff
@@ -590,10 +587,101 @@ void error() {
...
+    // Body below the limit, succeeds
+    addHttpResponse(200, exporter.exportResponse(100));
+    CompletableResultCode result =
+        exporter
+            .export(Collections.singletonList(generateFakeTelemetry()))
+            .join(10, TimeUnit.SECONDS);
+    assertThat(result.isSuccess()).isTrue();
+
+    // Body over the limit, fails with RESOURCE_EXHAUSTED
+    addHttpResponse(200, exporter.exportResponse(4 * 1024 * 1024 + 1));
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractHttpTelemetryExporterTest.java:606]: fixed in b86d92132

File: exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java [hunk 13]
```diff
@@ -332,6 +365,102 @@ void export() {
...
+    // Body below the limit, succeeds
+    addGrpcResponse(GrpcStatusCode.OK, null, exporter.exportResponse(100));
+    CompletableResultCode result =
+        exporter
+            .export(Collections.singletonList(generateFakeTelemetry()))
+            .join(10, TimeUnit.SECONDS);
+    assertThat(result.isSuccess()).isTrue();
+
+    // Body over the limit, fails with RESOURCE_EXHAUSTED
+    addGrpcResponse(GrpcStatusCode.OK, null, exporter.exportResponse(4 * 1024 * 1024 + 1));
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/otlp/testing-internal/src/main/java/io/opentelemetry/exporter/otlp/testing/internal/AbstractGrpcTelemetryExporterTest.java:384]: fixed in b86d92132

File: exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpGrpcSender.java [hunk 14]
```diff
@@ -165,58 +168,92 @@ public void onFailure(Call call, IOException e) {
...
+      // maxResponseBodySize (EOF is reached before the extra byte is read).
+      // If maxResponseBodySize is Long.MAX_VALUE, adding 1 would overflow. In that case use
+      // Long.MAX_VALUE directly — the overflow check can never trigger for such a large limit.
+      long readUpTo =
+          maxResponseBodySize == Long.MAX_VALUE ? Long.MAX_VALUE : maxResponseBodySize + 1;
+      Buffer wireBuffer = new Buffer();
+      try {
+        while (wireBuffer.size() <= maxResponseBodySize) {
+          long n = body.source().read(wireBuffer, readUpTo - wireBuffer.size());
+          if (n == -1L) {
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpGrpcSender.java:204]: fixed in b86d92132

**Comment by jkwatson** [comment]
I'm fine with the suggested API changes. 👍🏽

**Comment by jack-berg** [comment]
>  About the config. We don't really use the HttpSenderConfig I'll have to map it to something like: quarkus.otel.exporter.otlp.maxResponseBodySize.

I think the reason here is probably something to do with quarkus having its own configuration system and taking steps to ensure that only the quarkus vertx based sender is present, which cooperates with the quarkus config system.

But if its possible / useful to call the OTLP programmatic config API in quarkus, e.g. `OtlpHttpSpanExporter.builder().setEndpoint(..).setTimeout(..).setRetryPolicy(..).build()` (or whatever), then the config parameters programmatically set will manifest in getters on the `HttpSenderConfig` passed during sender initialization. The contract of the OTLP exporters is that the options you programmatically set are reflected in the config of the sender. I think it would be unintuitive if config parameters I explicitly hardcoded were ignored or overridden by something else. Just my 2 cents.

File: exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java [hunk 15]
```diff
@@ -319,24 +335,48 @@ private byte[] buf() {
...
+    // what an int can hold. In that case use Integer.MAX_VALUE — the overflow check can never
+    // trigger for such a large limit.
+    int readUpTo =
+        maxResponseBodySize >= Integer.MAX_VALUE
+            ? Integer.MAX_VALUE
+            : (int) (maxResponseBodySize + 1);
+
+    String contentEncoding = response.headers().firstValue("Content-Encoding").orElse(null);
+    if (contentEncoding != null && !"gzip".equalsIgnoreCase(contentEncoding)) {
+      throw new UnsupportedContentEncodingException(
```
Comment (jackshirazi) [inline exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java:354]: `identity` was added as now advertised being supported, but will be rejected here

File: exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpHttpSender.java [hunk 16]
```diff
@@ -141,39 +151,54 @@ public void onFailure(Call call, IOException e) {
...
+                        handleResponse(response, onResponse, onError);
                       }
                     }));
   }
 
+  private void handleResponse(
+      Response response, Consumer<HttpResponse> onResponse, Consumer<Throwable> onError) {
+    try (ResponseBody body = response.body()) {
+      String contentEncoding = response.header("Content-Encoding");
+      if (contentEncoding != null && !"gzip".equalsIgnoreCase(contentEncoding)) {
```
Comment (jackshirazi) [inline exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpHttpSender.java:163]: identity was added as now advertised being supported, but will be rejected here

File: exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java (hunk 15 above)
Comment (jack-berg) [reply exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java:354]: https://github.com/open-telemetry/opentelemetry-java/pull/8277

File: exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpHttpSender.java (hunk 16 above)
Comment (jack-berg) [reply exporters/sender/okhttp/src/main/java/io/opentelemetry/exporter/sender/okhttp/internal/OkHttpHttpSender.java:163]: https://github.com/open-telemetry/opentelemetry-java/pull/8277
